editor/src/clj/editor/lsp.clj

```diff
@@ -0,0 +1,440 @@
+
+(defn set-servers [new-servers]
+  (fn [{:keys [servers project] :as state}]
+    (let [old-servers (set (keys servers))
+          to-remove (set/difference old-servers new-servers)
+          to-add (set/difference new-servers old-servers)]
+      (as-> state $
+            (reduce (partial add-server! project) $ to-add)
+            (reduce remove-server! $ to-remove)))))
+
+(def ^:private dev (system/defold-dev?))
+(defonce ^:private running-lsps (when dev (atom {} :meta {:type ::running-lsps})))
+
+(defn make
+  "Create LSP manager that can be used for interacting with multiple LSP servers
+
+  Args:
+    project              the project node id
+    get-resource-node    fn of 3 args - project, resource and evaluation
+                         context - that should return current resource node id
+                         for the specified resource
+
+  Returns a channel that can be used to submit LSP manager commands (see other
+  public fns in this ns)"
+  [project get-resource-node]
+  (let [in (a/chan 128)]
+    (a/go-loop [state {:channels [in] ;; in + server output channels, for performance
+                       :project project
+                       :get-resource-node get-resource-node
+                       ;; servers is a map from server to server state, where
+                       ;; server is:
+                       ;; {:languages #{"lang"}
+                       ;;  :launcher {:command ["shell-command"]}}
+                       ;; and server state is:
+                       ;; {:in ch
+                       ;;  :out ch
+                       ;;  :status :initializing|:running|:down
+                       ;;  ; appears when diagnostics are published
+                       ;;  :diagnostics {resource [diagnostic]}
+                       ;;  ; appears on initialization
+                       ;;  :capabilities {:text-document-sync {:open-close boolean
+                       ;;                                      :change :none|:full|:incremental}}}
+                       :servers {}
+                       ;; map form server's :out ch to server, for performance
+                       :server-outputs {}
+                       ;; set of resources we monitor for changes
+                       :polled-resources #{}
+                       ;; map of resource to open view id
+                       :viewed-resources {}
+                       ;; inverse :viewed-resources, a map from view node id
+                       ;; to resource, used for performance
+                       :view-nodes {}
+                       ;; map {resource {:lines ["code..."] :version int}}
+                       ;; invariant: every viewed resource must be open
+                       :open-resources {}}]
+      (when dev (swap! running-lsps assoc in state))
+      (let [[value ch] (a/alts! (:channels state))]
+        (cond
+          ;; close the manager input => stop the manager
+          (and (nil? value) (= ch in))
```

Comment:
Just in case you have no prior experience with core.async, `nil` is special there: you can't put `nil` on a channel, and if the channel is closed, takes from it will return `nil`.